### PR TITLE
[RNTuple] Add test for compression algorithm

### PR DESCRIPTION
**Pull request by codetronaut**
Added the Test for the compression algorithm for RNTuple as a GSoC evaluation task.

### CODE REVIEW

File: tree/ntuple/v7/test/CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,7 @@ if(MSVC)
                                      ${CMAKE_CURRENT_BINARY_DIR}/libCustomStruct.dll)
 endif()
 ROOT_ADD_GTEST(ntuple ntuple.cxx LIBRARIES ROOTDataFrame ROOTNTuple MathCore CustomStruct)
+ROOT_ADD_GTEST(RNtuple RNtuple.cxx LIBRARIES ROOTDataFrame ROOTNTuple MathCore CustomStruct)
```
Comment (oshadura) [inline tree/ntuple/v7/test/CMakeLists.txt]: Please rename file in ntuple_compression.cxx, the same about name of executable.
Do you really need CustomStruct library as dependency?

File: tree/ntuple/v7/test/RNtuple.cxx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    TRandom3 rnd(42);
+
+    double chksumWrite = 0.0;
+    {
+        auto ntuple = RNTupleWriter::Recreate(std::move(model), "CompTest", file)
+
+        for (auto i = 0; i<NEvents ; ++i)
+        {
+            floodRand = rnd.Rndm();    
+            chksumWrite += floodRand;
```
Comment (oshadura) [inline tree/ntuple/v7/test/RNtuple.cxx]: Do you really need a checksunWrite? (are you using it somewhere?)

File: tree/ntuple/v7/test/RNtuple.cxx [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            ntuple->Fill();             
+        /*
```
Comment (oshadura) [inline tree/ntuple/v7/test/RNtuple.cxx]: Please remove this commented out code if you think it is not needed.

File: tree/ntuple/v7/test/RNtuple.cxx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        */
+        }
+            ntuple->write;
+            floodRand->clear();
+        }
+
+
+    auto compress = file->GetCompressionSettings();
+
+    ASSERT_EQ(compress, 101);
```
Comment (oshadura) [inline tree/ntuple/v7/test/RNtuple.cxx]: Are you sure that it is 101? I know RNtuple has different default settings.

**Comment by oshadura** [comment]
Other generic request: it is required to write meaningful commit messages and squash "non functional" changes (like remove one empty line and etc.).

**Comment by codetronaut** [comment]
Thanks, @oshadura, for the response I will try to correct the mistakes asap.

File: tree/ntuple/v7/test/ntuple_compression.cxx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        ntuple->Fill();    
+    }
+    
+    auto compress = (*file)->GetCompressionSettings();
+        
+    ASSERT_EQ(compress, 404);
+    
+    file->Close();
+    delete file;
+}
```
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_compression.cxx:72]: Missing trailing newline

File: tree/ntuple/v7/test/ntuple_compression.cxx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+TEST(RNTuple, nTupleCompressionTest){
+    
+
+    FileRaii fileGuard("test_ntuple_compression.root");
+    auto file = TFile::Open(fileGuard.GetPath().c_str(), "RECREATE");
+
+    auto model = RNTupleModel::Create();
+    
+    const int NEvents = 1000;
```
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_compression.cxx:52]: The test would also work with just 1 event, wouldn't it?

File: tree/ntuple/v7/test/ntuple_compression.cxx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+   FileRaii(const FileRaii&) = delete;
+   FileRaii& operator=(const FileRaii&) = delete;
+   ~FileRaii() { std::remove(fPath.c_str()); }
+   std::string GetPath() const { return fPath; }
+};
+
+}
+
+TEST(RNTuple, nTupleCompressionTest){
+
```
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_compression.cxx:45]: Remove empty lines

File: tree/ntuple/v7/test/ntuple_compression.cxx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    auto model = RNTupleModel::Create();
+    
+    const int NEvents = 1000;
+
+    auto& randData = *model->MakeField<std::vector<double>>("data");
+   
+    auto ntuple = RNTupleWriter::Recreate(std::move(model), "CompTest", fileGuard.GetPath().c_str(), RNTupleWriteOptions());
+    
+    TRandom3 rnd(42);
```
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_compression.cxx:58]: We don't need random data, I think

File: tree/ntuple/v7/test/ntuple_compression.cxx [hunk 9]
```diff
@@ -0,0 +1,72 @@
+#include <ROOT/RNTuple.hxx>
```
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_compression.cxx:1]: I think a number of includes are not necessary.  Please trim down to the minimum includes list.

File: tree/ntuple/v7/test/CMakeLists.txt [hunk 10]
```diff
@@ -22,6 +22,7 @@ if(MSVC)
                                      ${CMAKE_CURRENT_BINARY_DIR}/libCustomStruct.dll)
 endif()
 ROOT_ADD_GTEST(ntuple ntuple.cxx LIBRARIES ROOTDataFrame ROOTNTuple MathCore CustomStruct)
+ROOT_ADD_GTEST(ntuple_compression ntuple_compression.cxx LIBRARIES ROOTDataFrame ROOTNTuple MathCore)
```
Comment (jblomer) [inline tree/ntuple/v7/test/CMakeLists.txt:25]: If `TRandom` is not used, the `MathCore` dependency can be removed. The data frame dependency seem to be avoidable, too.

**Comment by codetronaut** [comment]
> Many thanks! More unit tests are always a good thing!
> 
> For this particular one, I am not quite sure what it tests wrt. to RNTuple. If I understand correctly, it checks whether a newly created ROOT file uses 404 as its default compression settings, no?

I Will work on all the issues mentioned asap.
Thanks for your review.

**Comment by codetronaut** [comment]
> Many thanks! More unit tests are always a good thing!
> 
> For this particular one, I am not quite sure what it tests wrt. to RNTuple. If I understand correctly, it checks whether a newly created ROOT file uses 404 as its default compression settings, no?

hello @jblomer ,
so I have to first fill the file with data and compress(using RNTupleZip) it and then compare it with the original size when it was initially filled?

**Comment by jblomer** [comment]
> so I have to first fill the file with data and compress(using RNTupleZip) it and then compare it with the original size when it was initially filled?

I'm still a bit puzzled. Can you please describe which functionality or property of RNTuple you'd like to test?

**Comment by codetronaut** [comment]
> > so I have to first fill the file with data and compress(using RNTupleZip) it and then compare it with the original size when it was initially filled?
> 
> I'm still a bit puzzled. Can you please describe which functionality or property of RNTuple you'd like to test?

hello,
what I intend to do is to create an RNTuple, populate it with random data, and checking default compression settings used for generated files (by default, without explicitly saying which compression algorithm I want to use).

**Comment by jblomer** [comment]
Ok, I see.  In this case, you would not need to generate a lot of data, right?  A more concise test would be to create an empty RNTuple, to add one value, and then to open it and check what's the compression setting.

**Comment by codetronaut** [comment]
> Ok, I see.  In this case, you would not need to generate a lot of data, right?  A more concise test would be to create an empty RNTuple, to add one value, and then to open it and check what's the compression setting.

Yes, and I'm looking in https://github.com/root-project/root/blob/master/tree/ntuple/v7/test/ntuple_zip.cxx#L67

Also discussed my approach in previous comment, please tell me am I going right ?

File: tree/ntuple/v7/test/ntuple_compression.cxx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    
+    TRandom3 rnd(42);
+    
+    for (auto i=0; i<NEvents; ++i)
+    {
+        randData = rnd.Rndm();
+        ntuple->Fill();    
+    }
+    
+    auto compress = (*file)->GetCompressionSettings();
```
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_compression.cxx:66]: This should probably check the _ntuple_'s compression algorithm, which in general does not need to be the same as the one in the file.

**Comment by jblomer** [comment]
I think the PR goes into the right direction.  Please use just 1 simple event instead of 1000, because we are really testing only if the compression setting gets applied.  The `ntuple_zip` unit tests are _not_ really relevant here -- they test the compression routines themselves and not if the correct compression settings were used.

**Comment by jblomer** [comment]
@codetronaut Just checking if you are still working on this.

**Comment by codetronaut** [comment]
@jblomer , sorry to say but I am not able to find time for this, right now I am closing this issue but will open as soon as possible when I'm ready to work on this.
